File: parser/parser.go
package parser

import (
	"errors"
	"strconv"
)

type Parser struct {
	current int
	tokens  []Token
	errors  []Error
}

func CreateParser(tokens []Token) *Parser {
	parser := &Parser{
		current: 0,
		tokens:  tokens,
		errors:  make([]Error, 0),
	}

	return parser
}

func (parser *Parser) Parse() Syntax {
	return parser.parseProgram()
}

func (parser *Parser) parseProgram() ProgramSyntax {
	var statements []Syntax
	for !parser.isAtEnd() {
		statements = append(statements, parser.parseStatement())
	}

	return ProgramSyntax{
		Statements: statements,
	}
}

func (parser *Parser) parseStatement() Syntax {
	var nextType = parser.read().tokenType
	switch nextType {
	case TokenTypeInputKeyword:
		return parser.parseInputStatement()
	case TokenTypeOutputKeyword:
		return parser.parseOutputStatement()
	case TokenTypeVariableKeyword:
		return parser.parseVariableStatement()
	case TokenTypeResourceKeyword:
		return parser.parseResourceStatement()
	}

	parser.emitError("Unexpected token")
	return nil
}

func (parser *Parser) parseInputStatement() Syntax {
	return parser.doWithRecovery(parser.current-1, TokenTypeSemicolon, func() (Syntax, error) {
		keyword := parser.prev()
		inputType := parser.parseIdentifier()
		identifier := parser.parseIdentifier()
		semicolon := parser.expect(TokenTypeSemicolon)

		return InputDeclSyntax{
			InputKeyword: keyword,
			Type:         inputType,
			Identifier:   identifier,
			Semicolon:    semicolon,
		}, nil
	})
}

func (parser *Parser) parseOutputStatement() Syntax {
	return parser.doWithRecovery(parser.current-1, TokenTypeSemicolon, func() (Syntax, error) {
		keyword := parser.prev()
		identifier := parser.parseIdentifier()
		colon := parser.expect(TokenTypeColon)
		expression, err := parser.parseExpression()
		if err != nil {
			return nil, err
		}
		semicolon := parser.expect(TokenTypeSemicolon)

		return OutputDeclSyntax{
			OutputKeyword: keyword,
			Identifier:    identifier,
			Colon:         colon,
			Expression:    expression,
			Semicolon:     semicolon,
		}, nil
	})
}

func (parser *Parser) parseVariableStatement() Syntax {
	return parser.doWithRecovery(parser.current-1, TokenTypeSemicolon, func() (Syntax, error) {
		keyword := parser.prev()
		identifier := parser.parseIdentifier()
		colon := parser.expect(TokenTypeColon)
		expression, err := parser.parseExpression()
		if err != nil {
			return nil, err
		}
		semicolon := parser.expect(TokenTypeSemicolon)

		return VarDeclSyntax{
			VariableKeyword: keyword,
			Identifier:      identifier,
			Colon:           colon,
			Expression:      expression,
			Semicolon:       semicolon,
		}, nil
	})
}

func (parser *Parser) parseResourceStatement() Syntax {
	return parser.doWithRecovery(parser.current-1, TokenTypeSemicolon, func() (Syntax, error) {
		keyword := parser.prev()
		provider := parser.parseIdentifier()
		typeString := parser.parseString()
		identifier := parser.parseIdentifier()
		colon := parser.expect(TokenTypeColon)
		expression, err := parser.parseExpression()
		if err != nil {
			return nil, err
		}
		semicolon := parser.expect(TokenTypeSemicolon)

		return ResourceDeclSyntax{
			ResourceKeyword: keyword,
			Provider:        provider,
			Type:            typeString,
			Identifier:      identifier,
			Colon:           colon,
			Expression:      expression,
			Semicolon:       semicolon,
		}, nil
	})
}

func (parser *Parser) parseString() StringSyntax {
	stringToken := parser.expect(TokenTypeString)

	return StringSyntax{
		StringToken: stringToken,
	}
}

func (parser *Parser) parseIdentifier() IdentifierSyntax {
	identifier := parser.expect(TokenTypeIdentifier)

	return IdentifierSyntax{
		Identifier: identifier,
	}
}

func (parser *Parser) parseExpression() (Syntax, error) {
	return parser.parseOr()
}

func (parser *Parser) parseOr() (Syntax, error) {
	expression, err := parser.parseAnd()
	if err != nil {
		return nil, err
	}

	if parser.match(TokenTypeBinaryOr) {
		var operator = parser.prev()
		rightExpression, err := parser.parseOr()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpOr,
		}, nil
	}

	return expression, nil
}

func (parser *Parser) parseAnd() (Syntax, error) {
	expression, err := parser.parseEquality()
	if err != nil {
		return nil, err
	}

	if parser.match(TokenTypeBinaryOr) {
		var operator = parser.prev()
		rightExpression, err := parser.parseAnd()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpAnd,
		}, nil
	}

	return expression, nil
}

func (parser *Parser) parseEquality() (Syntax, error) {
	expression, err := parser.parseComparison()
	if err != nil {
		return nil, err
	}

	if parser.match(TokenTypeEquals) {
		var operator = parser.prev()
		rightExpression, err := parser.parseEquality()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpEquals,
		}, nil
	}
	if parser.match(TokenTypeNotEquals) {
		var operator = parser.prev()
		rightExpression, err := parser.parseEquality()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpNotEquals,
		}, nil
	}
	if parser.match(TokenTypeEqualsInsensitive) {
		var operator = parser.prev()
		rightExpression, err := parser.parseEquality()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpEqualsInsensitive,
		}, nil
	}
	if parser.match(TokenTypeNotEqualsInsensitive) {
		var operator = parser.prev()
		rightExpression, err := parser.parseEquality()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpNotEqualsInsensitive,
		}, nil
	}

	return expression, nil
}

func (parser *Parser) parseComparison() (Syntax, error) {
	expression, err := parser.parseAddition()
	if err != nil {
		return nil, err
	}

	if parser.match(TokenTypeGreaterThan) {
		var operator = parser.prev()
		rightExpression, err := parser.parseComparison()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpGreaterThan,
		}, nil
	}
	if parser.match(TokenTypeGreaterThanOrEqual) {
		var operator = parser.prev()
		rightExpression, err := parser.parseComparison()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpGreaterThanOrEqual,
		}, nil
	}
	if parser.match(TokenTypeLessThan) {
		var operator = parser.prev()
		rightExpression, err := parser.parseComparison()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpLessThan,
		}, nil
	}
	if parser.match(TokenTypeLessThanOrEqual) {
		var operator = parser.prev()
		rightExpression, err := parser.parseComparison()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpLessThanOrEqual,
		}, nil
	}

	return expression, nil
}

func (parser *Parser) parseAddition() (Syntax, error) {
	expression, err := parser.parseMultiplication()
	if err != nil {
		return nil, err
	}

	if parser.match(TokenTypePlus) {
		var operator = parser.prev()
		rightExpression, err := parser.parseAddition()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpAdd,
		}, nil
	}
	if parser.match(TokenTypeMinus) {
		var operator = parser.prev()
		rightExpression, err := parser.parseAddition()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpSubtract,
		}, nil
	}

	return expression, nil
}

func (parser *Parser) parseMultiplication() (Syntax, error) {
	expression, err := parser.parseUnary()
	if err != nil {
		return nil, err
	}

	if parser.match(TokenTypeAsterisk) {
		var operator = parser.prev()
		rightExpression, err := parser.parseMultiplication()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpMultiply,
		}, nil
	}
	if parser.match(TokenTypeSlash) {
		var operator = parser.prev()
		rightExpression, err := parser.parseMultiplication()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpDivide,
		}, nil
	}
	if parser.match(TokenTypeModulus) {
		var operator = parser.prev()
		rightExpression, err := parser.parseMultiplication()
		if err != nil {
			return nil, err
		}
		return BinaryOperationSyntax{
			LeftExpression:  expression,
			Operator:        operator,
			RightExpression: rightExpression,
			Operation:       BinaryOpModulus,
		}, nil
	}

	return expression, nil
}

func (parser *Parser) parseUnary() (Syntax, error) {
	if parser.match(TokenTypeExclamation) {
		var operator = parser.prev()
		expression, err := parser.parseUnary()
		if err != nil {
			return nil, err
		}
		return UnaryOperationSyntax{
			Operator:   operator,
			Expression: expression,
			Operation:  UnaryOpNot,
		}, err
	}
	if parser.match(TokenTypeMinus) {
		var operator = parser.prev()
		expression, err := parser.parseUnary()
		if err != nil {
			return nil, err
		}
		return UnaryOperationSyntax{
			Operator:   operator,
			Expression: expression,
			Operation:  UnaryOpNegate,
		}, nil
	}

	return parser.parseMemberAccess()
}

func (parser *Parser) parseMemberAccess() (Syntax, error) {
	expression, err := parser.parseFunctionCall()
	if err != nil {
		return nil, err
	}

	for true {
		if parser.match(TokenTypeDot) {
			dot := parser.prev()
			member, err := parser.parseFunctionCall()
			if err != nil {
				return nil, err
			}

			expression = PropertyAccessSyntax{
				Parent:   expression,
				Dot:      dot,
				Property: member,
			}
		} else if parser.match(TokenTypeLeftSquare) {
			leftSquare := parser.prev()
			member, err := parser.parseExpression()
			if err != nil {
				return nil, err
			}
			rightSquare := parser.expect(TokenTypeRightSquare)

			expression = ArrayAccessSyntax{
				Parent:      expression,
				LeftSquare:  leftSquare,
				Property:    member,
				RightSquare: rightSquare,
			}
		} else if parser.isAtEnd() {
			parser.emitError("Unexpected end")
		} else {
			break
		}
	}

	return expression, nil
}

func (parser *Parser) parseFunctionCall() (Syntax, error) {
	primary, err := parser.parsePrimary()
	if err != nil {
		return nil, err
	}

	if parser.match(TokenTypeLeftParen) {
		leftParen := parser.prev()
		var arguments []Syntax

		for !parser.match(TokenTypeRightParen) {
			if parser.isAtEnd() {
				parser.emitError("Unexpected end")
				return nil, errors.New("Unexpected end")
			}

			if len(arguments) > 0 {
				parser.expect(TokenTypeComma)
			}

			argument, err := parser.parseExpression()
			if err != nil {
				return nil, err
			}
			arguments = append(arguments, argument)
		}

		rightParen := parser.prev()

		return FunctionCallSyntax{
			Parent:     primary,
			LeftParen:  leftParen,
			Arguments:  arguments,
			RightParen: rightParen,
		}, nil
	}

	return primary, nil
}

func (parser *Parser) parsePrimary() (Syntax, error) {
	if parser.match(TokenTypeNumber) {
		literal := parser.prev()
		value, _ := strconv.ParseInt(literal.RawValue, 10, 64) // todo handle errors
		return NumericLiteralSyntax{
			Literal: literal,
			Value:   int(value), // todo avoid this cast
		}, nil
	}
	if parser.match(TokenTypeString) {
		stringToken := parser.prev()
		return StringSyntax{
			StringToken: stringToken,
		}, nil
	}
	if parser.match(TokenTypeIdentifier) {
		identifier := parser.prev()
		return IdentifierSyntax{
			Identifier: identifier,
		}, nil
	}
	if parser.match(TokenTypeFalseKeyword) {
		literal := parser.prev()
		return BooleanLiteralSyntax{
			Literal: literal,
			Value:   false,
		}, nil
	}
	if parser.match(TokenTypeTrueKeyword) {
		literal := parser.prev()
		return BooleanLiteralSyntax{
			Literal: literal,
			Value:   true,
		}, nil
	}
	if parser.match(TokenTypeNullKeyword) {
		literal := parser.prev()
		return NullLiteralSyntax{
			Literal: literal,
		}, nil
	}
	if parser.match(TokenTypeLeftParen) {
		leftParen := parser.prev()
		expression, err := parser.parseExpression()
		if err != nil {
			return nil, err
		}
		rightParen := parser.expect(TokenTypeRightParen)
		return GroupingSyntax{
			LeftParen:  leftParen,
			Expression: expression,
			RightParen: rightParen,
		}, nil
	}
	if parser.match(TokenTypeLeftBrace) {
		return parser.parseObject()
	}
	if parser.match(TokenTypeLeftSquare) {
		return parser.parseArray()
	}

	parser.emitError("Unexpected token")
	return nil, errors.New("Unexpected token")
}

func (parser *Parser) parseObject() (*ObjectSyntax, error) {
	leftBrace := parser.prev()
	var objectProperties []ObjectPropertySyntax

	for !parser.match(TokenTypeRightBrace) {
		if parser.isAtEnd() {
			parser.emitError("Unexpected end")
		}

		identifier := parser.parseIdentifier()
		colon := parser.expect(TokenTypeColon)
		expression, err := parser.parseExpression()
		if err != nil {
			return nil, err
		}

		var objectProperty = ObjectPropertySyntax{
			Identifier: identifier,
			Colon:      colon,
			Expression: expression,
		}
		objectProperties = append(objectProperties, objectProperty)

		// either expect a comma or the end of the object (since a trailing comma is optional).
		// don't use match for '}' as we want the outer loop to pick it up.
		if !parser.match(TokenTypeComma) && !parser.check(TokenTypeRightBrace) {
			if !parser.check(TokenTypeRightBrace) {
				parser.emitError("Unexpected character")
			}
		}
	}

	rightBrace := parser.prev()

	return &ObjectSyntax{
		LeftBrace:  leftBrace,
		Properties: objectProperties,
		RightBrace: rightBrace,
	}, nil
}

func (parser *Parser) parseArray() (*ArraySyntax, error) {
	leftSquare := parser.prev()
	var items []Syntax

	for !parser.match(TokenTypeRightSquare) {
		if parser.isAtEnd() {
			parser.emitError("Unexpected end")
		}

		expression, err := parser.parseExpression()
		if err != nil {
			return nil, err
		}
		items = append(items, expression)

		// either expect a comma or the end of the object (since a trailing comma is optional).
		// don't use match for ']' as we want the outer loop to pick it up.
		if !parser.match(TokenTypeComma) && !parser.check(TokenTypeRightSquare) {
			if !parser.check(TokenTypeRightSquare) {
				parser.emitError("Unexpected character")
			}
		}
	}

	rightSquare := parser.prev()

	return &ArraySyntax{
		LeftSquare:  leftSquare,
		Items:       items,
		RightSquare: rightSquare,
	}, nil
}

func (parser *Parser) getPosition() Position {
	return Position{
		start: parser.tokens[parser.current].position.start,
		end:   parser.tokens[parser.current].position.end,
	}
}

func (parser *Parser) emitError(message string) {
	err := Error{
		message:  message,
		position: parser.getPosition(),
	}

	parser.errors = append(parser.errors, err)
}

func (parser *Parser) isAtEnd() bool {
	return parser.current >= len(parser.tokens) || parser.peek().tokenType == TokenTypeEndOfFile
}

func (parser *Parser) peek() Token {
	return parser.tokens[parser.current]
}

func (parser *Parser) read() Token {
	output := parser.peek()
	parser.current++
	return output
}

func (parser *Parser) prev() Token {
	return parser.tokens[parser.current-1]
}

func (parser *Parser) expect(tokenType TokenType) Token {
	token := parser.read()
	if token.tokenType == tokenType {
		return token
	}

	parser.emitError("Expected ...")
	return token //todo remove this
}

func (parser *Parser) match(tokenType TokenType) bool {
	if parser.check(tokenType) {
		parser.read()
		return true
	}

	return false
}

func (parser *Parser) check(tokenType TokenType) bool {
	if parser.isAtEnd() {
		return false
	}

	return parser.peek().tokenType == tokenType
}

func (parser *Parser) doWithRecovery(start int, matchTokenType TokenType, innerFunc func() (Syntax, error)) Syntax {
	output, err := innerFunc()
	if err == nil {
		return output
	}

	for !parser.isAtEnd() {
		if parser.match(matchTokenType) {
			break
		}
		parser.read()
	}

	return ErrorSyntax{
		Tokens: parser.tokens[start:parser.current],
	}
}

File: src/Bicep/Parser/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Bicep.Syntax;

namespace Bicep.Parser
{
    public class Parser
    {
        private readonly TokenReader reader;
        private readonly IList<Error> errors = new List<Error>();

        public Parser(IEnumerable<Token> tokens)
        {
            this.reader = new TokenReader(tokens);
        }

        public SyntaxBase Parse()
            => Program();

        private SyntaxBase Program()
        {
            var statements = new List<SyntaxBase>();
            while (!IsAtEnd())
            {
                var statement = Statement();
                statements.Add(statement);
            }

            var endOfFile = reader.Read();

            return new ProgramSyntax(statements, endOfFile);
        }

        private SyntaxBase Statement()
        {
            var nextType = reader.Read().Type;
            switch (nextType)
            {
                case TokenType.InputKeyword:
                    return InputStatement();
                case TokenType.OutputKeyword:
                    return OutputStatement();
                case TokenType.VariableKeyword:
                    return VariableStatement();
                case TokenType.ResourceKeyword:
                    return ResourceStatement();
            }

            throw new NotImplementedException();
        }

        private SyntaxBase InputStatement()
            => WithRecovery(() => {
                var keyword = reader.Prev();
                var type = Identifier();
                var identifier = Identifier();
                var semicolon = Expect(TokenType.Semicolon, "");

                return new InputDeclSyntax(keyword, type, identifier, semicolon);
            });

        private SyntaxBase OutputStatement()
            => WithRecovery(() => {
                var keyword = reader.Prev();
                var identifier = Identifier();
                var colon = Expect(TokenType.Colon, "");
                var expression = Expression();
                var semicolon = Expect(TokenType.Semicolon, "");

                return new OutputDeclSyntax(keyword, identifier, colon, expression, semicolon);
            });

        private SyntaxBase VariableStatement()
            => WithRecovery(() => {
                var keyword = reader.Prev();
                var identifier = Identifier();
                var colon = Expect(TokenType.Colon, "");
                var expression = Expression();
                var semicolon = Expect(TokenType.Semicolon, "");

                return new VarDeclSyntax(keyword, identifier, colon, expression, semicolon);
            });

        private SyntaxBase ResourceStatement()
            => WithRecovery(() => {
                var keyword = reader.Prev();
                var provider = Identifier();
                var type = String();
                var identifier = Identifier();
                var colon = Expect(TokenType.Colon, "");
                var expression = Expression();
                var semicolon = Expect(TokenType.Semicolon, "");

                return new ResourceDeclSyntax(keyword, provider, type, identifier, colon, expression, semicolon);
            });

        private StringSyntax String()
        {
            var stringToken = Expect(TokenType.String, "...");

            return new StringSyntax(stringToken);
        }

        private IdentifierSyntax Identifier()
        {
            var identifier = Expect(TokenType.Identifier, "...");

            return new IdentifierSyntax(identifier);
        }

        private SyntaxBase Expression()
            => Or();

        private SyntaxBase Or()
        {
            var expression = And();

            if (Match(TokenType.BinaryOr))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Or(), BinaryOperation.Or);
            }

            return expression;
        }

        private SyntaxBase And()
        {
            var expression = Equality();

            if (Match(TokenType.BinaryAnd))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, And(), BinaryOperation.And);
            }

            return expression;
        }

        private SyntaxBase Equality()
        {
            var expression = Comparison();

            if (Match(TokenType.Equals))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Equality(), BinaryOperation.Equals);
            }

            if (Match(TokenType.NotEquals))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Equality(), BinaryOperation.NotEquals);
            }

            if (Match(TokenType.EqualsInsensitive))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Equality(), BinaryOperation.EqualsInsensitive);
            }

            if (Match(TokenType.NotEqualsInsensitive))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Equality(), BinaryOperation.NotEqualsInsensitive);
            }

            return expression;
        }

        private SyntaxBase Comparison()
        {
            var expression = Addition();

            if (Match(TokenType.GreaterThan))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Comparison(), BinaryOperation.GreaterThan);
            }

            if (Match(TokenType.GreaterThanOrEqual))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Comparison(), BinaryOperation.GreaterThanOrEqual);
            }

            if (Match(TokenType.LessThan))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Comparison(), BinaryOperation.LessThan);
            }

            if (Match(TokenType.LessThanOrEqual))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Comparison(), BinaryOperation.LessThanOrEqual);
            }

            return expression;
        }

        private SyntaxBase Addition()
        {
            var expression = Multiplication();

            if (Match(TokenType.Plus))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Addition(), BinaryOperation.Add);
            }

            if (Match(TokenType.Minus))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Addition(), BinaryOperation.Subtract);
            }

            return expression;
        }

        private SyntaxBase Multiplication()
        {
            var expression = Unary();

            if (Match(TokenType.Asterisk))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Multiplication(), BinaryOperation.Multiply);
            }

            if (Match(TokenType.Slash))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Multiplication(), BinaryOperation.Divide);
            }

            if (Match(TokenType.Modulus))
            {
                var operatorToken = reader.Prev();
                return new BinaryOperationSyntax(expression, operatorToken, Multiplication(), BinaryOperation.Modulus);
            }

            return expression;
        }

        private SyntaxBase Unary()
        {
            if (Match(TokenType.Exclamation))
            {
                var operatorToken = reader.Prev();
                return new UnaryOperationSyntax(operatorToken, Unary());
            }

            if (Match(TokenType.Minus))
            {
                var operatorToken = reader.Prev();
                return new UnaryOperationSyntax(operatorToken, Unary());
            }

            return MemberAccess();
        }

        private SyntaxBase MemberAccess()
        {
            var output = FunctionCall();

            while (true)
            {
                if (Match(TokenType.Dot))
                {
                    var dot = reader.Prev();
                    var member = FunctionCall();

                    output = new PropertyAccessSyntax(output, dot, member);
                }
                else if (Match(TokenType.LeftSquare))
                {
                    var leftSquare = reader.Prev();
                    var member = Expression();
                    var rightSquare = Expect(TokenType.RightSquare, "");

                    output = new ArrayAccessSyntax(output, leftSquare, member, rightSquare);
                }
                else if (IsAtEnd())
                {
                    throw new Exception("");
                }
                else
                {
                    break;
                }
            }

            return output;
        }

        private SyntaxBase FunctionCall()
        {
            var primary = Primary();

            if (Match(TokenType.LeftParen))
            {
                var leftParen = reader.Prev();
                var arguments = new List<SyntaxBase>();
                var separators = new List<Token>();
                while (!Match(TokenType.RightParen))
                {
                    if (IsAtEnd())
                    {
                        throw new Exception("");
                    }

                    if (arguments.Any())
                    {
                        var comma = Expect(TokenType.Comma, "");
                        separators.Add(comma);
                    }
    
                    var argument = Expression();
                    arguments.Add(argument);
                }
                var rightParen = reader.Prev();

                var syntaxList = new SeparatedSyntaxList(arguments, separators, TextSpan.BetweenNonInclusive(leftParen, rightParen));
                return new FunctionCallSyntax(primary, leftParen, syntaxList, rightParen);
            }

            return primary;
        }

        private SyntaxBase Primary()
        {
            if (Match(TokenType.Number))
            {
                // todo parse ints safely
                var literal = reader.Prev();
                var value = int.Parse(literal.Text);
                return new NumericLiteralSyntax(literal, value);
            }

            if (Match(TokenType.String))
            {
                // todo string interpolation
                var stringToken = reader.Prev();
                return new StringSyntax(stringToken);
            }

            if (Match(TokenType.Identifier))
            {
                var identifier = reader.Prev();
                return new IdentifierSyntax(identifier);
            }

            if (Match(TokenType.FalseKeyword))
            {
                var literal = reader.Prev();
                return new BooleanLiteralSyntax(literal, false);
            }

            if (Match(TokenType.TrueKeyword))
            {
                var literal = reader.Prev();
                return new BooleanLiteralSyntax(literal, true);
            }

            if (Match(TokenType.NullKeyword))
            {
                var literal = reader.Prev();
                return new NullLiteralSyntax(literal);
            }

            if (Match(TokenType.LeftParen))
            {
                var openParen = reader.Prev();
                var expression = Expression();
                var closeParen = Expect(TokenType.RightParen, "");

                return new GroupingSyntax(openParen, expression, closeParen);
            }

            if (Match(TokenType.LeftBrace))
            {
                return Object();
            }

            if (Match(TokenType.LeftSquare))
            {
                return Array();
            }

            throw new Exception("");
        }

        private ObjectSyntax Object()
        {
            var openBrace = reader.Prev();
            var properties = new List<ObjectPropertySyntax>();
            var separators = new List<Token>();
            while (!Match(TokenType.RightBrace))
            {
                if (IsAtEnd())
                {
                    throw new Exception("");
                }

                var identifier = Identifier();
                var colon = Expect(TokenType.Colon, "");
                var expression = Expression();

                var objectProperty = new ObjectPropertySyntax(identifier, colon, expression);
                properties.Add(objectProperty);

                if (Match(TokenType.Comma))
                {
                    var comma = reader.Prev();
                    separators.Add(comma);
                }
                else
                {
                    Expect(TokenType.RightBrace, "");
                    break;
                }
            }
            var closeBrace = reader.Prev();

            var syntaxList = new SeparatedSyntaxList(properties, separators, TextSpan.BetweenNonInclusive(openBrace, closeBrace));
            return new ObjectSyntax(openBrace, syntaxList, closeBrace);
        }

        private ArraySyntax Array()
        {
            var openSquare = reader.Prev();
            var items = new List<SyntaxBase>();
            var separators = new List<Token>();
            while (!Match(TokenType.RightSquare))
            {
                if (IsAtEnd())
                {
                    throw new Exception("");
                }

                var expression = Expression();
                items.Add(expression);

                if (Match(TokenType.Comma))
                {
                    var comma = reader.Prev();
                    separators.Add(comma);
                }
                else
                {
                    Expect(TokenType.RightSquare, "");
                    break;
                }
            }
            var closeSquare = reader.Prev();

            var syntaxList = new SeparatedSyntaxList(items, separators, TextSpan.BetweenNonInclusive(openSquare, closeSquare));
            return new ArraySyntax(openSquare, syntaxList, closeSquare);
        }

        private SyntaxBase WithRecovery<TSyntax>(Func<TSyntax> syntaxFunc)
            where TSyntax : SyntaxBase
        {
            return syntaxFunc();
            /* TODO recovery
            try
            {
                return syntaxFunc();
            }
            catch (Exception exception)
            {
                
            }
            */
        }

        private bool IsAtEnd()
        {
            return reader.IsAtEnd() || reader.Peek().Type == TokenType.EndOfFile;
        }

        private Token Expect(TokenType type, string message)
        {
            var token = reader.Read();
            if (token.Type == type)
            {
                return token;
            }

            throw new Exception(message);
        }

        private bool Match(TokenType type)
        {
            if (Check(type))
            {
                reader.Read();
                return true;
            }

            return false;
        }

        private bool Check(TokenType type)
        {
            if (IsAtEnd())
            {
                return false;
            }

            return reader.Peek().Type == type;
        }
    }
}